### PR TITLE
68 model with covariates

### LINKED ISSUE #68: model with covariates
When covariates are ready, we can test them in our base model of total population. This will require some minor revisions to the model, which I will start working on now.

### LINKED ISSUE #85: overestimating detection rates for Russian controlled oblasts
[Russia blocked access to Facebook and Instagram](https://netblocks.org/reports/twitter-and-facebook-restricted-in-russia-amid-conflict-with-ukraine-JBZrogB6) soon after their invasion of Ukraine, and this likely also affected Ukrainian territory under Russian control (Crimea, Sevastopol, Donetsk, Luhansk). 

The model is currently not able to detect this effect on the Facebook and Instagram detection rates. I have tried including covariates measuring areas under Russian control, but this does not seem to be working. The model estimates a sharp decline in population rather than a sharp decline in detection rates (see example in figure below). 

I would like to explore solutions to this problem that my include: 
- [x] 1. informative priors on population sizes periodically through the time series when external data can help inform this.
- [x] 2. modifying the regression structure on detection rates so that Facebook and Instagram detection rates are estimated by the same regression
- [x] 3. excluding Russian-controlled oblasts from the observation models but keep them in the process model
- [x] 4. include baseline detection rate for each oblast as a covariate

image

### PR DESCRIPTION
**Pull request by doug-leasure**
## Issues addressed

Closes #68 
Closes #85

## Summary of changes  

This pull request includes a model `./src/2_model/models/3_covs_model.stan` with the following characteristics:  
1. Allows time- and location-varying covariates on population growth rates in the process model
2. Allows time- and location-varying covariates on detection rates in the observation model (although, I would avoid putting covariates here if possible)  
3. Includes random effects of time and location on detection rates in the observation model
4. Covariates are selected by editing input files: `./wd/in/ua_r_covariates_oblast_selection.csv` (for process model covariates) or `./wd/in/ua_p_covariates_oblast_selection.csv` (for observation model covariates).  These files were based on the covariates description file `./wd/out/covariates/final/ua_covariates_oblast_description.csv` with a "select" column added to it.
6. Incorporates COD-PS from mid-2023 as empricial prior to constrain the population time series on that date with adjustable level of measurement error
7. Allows individual locations to be excluded from the observation model, the process model, or both (i.e. to exclude low-data quality locations from the likelihoods in the observation models while still including them in the process model),
8. Manually calculates the case-wise likelihood required for LOO cross-validation. 

The pull request also includes a model `./src/2_model/2_props_model.stan` that uses the proportional disaggregation approach, but does not yet include covariates. This is a modified version of this model that aligns it with the covariates model.

The scripts `./src/2_model/1_mcmc.R` and `.../2_eval.R` were improved to accommodate these different models, producing similar diagnostics and plots. 

## Suggestions for review

There is a lot here, so please do not feel like you need to review every detail. If you would like to modify anything, feel free to commit changes to this branch to include them with the pull request.

@njirons , please have a look at the `2_props_model` and feel free to adjust as you see fit. The MCMC chains were not perfectly behaved for this one after I modified your props model to harmonise it with my other covs models. 

@edarin, please take a look from a covariates perspective and let me know if you see any opportunities to stream line. Also, let's make sure that the outputs from this model align with the dashboard requirements. You can find the outputs in `K:/DemSci/Projects/2023_WHO_Ukraine_Population/wd/wd_doug/out/modelling/3_covs_model/mcmc/fit_3_covs_model.rds`.

@alaparicioc , please take a look at these models in comparison to other models you are working on and let me know if you see opportunities to harmonise with them or learn from each other.

## Input data
All of the input files are defined in the call to `model_data()` in the script `./src/2_model/1_mcmc.R`.  I put all of the required inputs into `K:/DemSci/Projects/2023_WHO_Ukraine_Population/wd/wd_doug/`.

Note: It should be possible to recreate all inputs from the covariates pipeline in the repository, except for data that are stored in `in_dir`. Those are files that are either sensitive (i.e. should not be put on GitHub) or should be customised locally for individual model runs.  

## Covariates - Proportions model

I did not include covariates in the proportions model, although I am considering doing that step before finalising this pull request. Alternatively, we could save that for later if needed.

### CODE REVIEW

**Review by njirons — APPROVED** [review]
20_props_model.stan ready for action

File: src/2_model/alt/models_doug/2_props_model/20_props_model.stan [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+  vector[T * I] log_p_G;
+  vector[n_F + n_G] log_lik;
+  vector[I-1] logit_pi0;
+
+  // population process model
+  pi[1] = N0/y_N_tot[1];
+  N[ti_N0] = y_N_tot[1] * pi[1]; 
+  logit_pi0 = log(pi[1,1:(I-1)]/(1-pi[1,I]));
+  for (t in 2 : T) {
+    pi[t] = softmax(append_row(logit_pi[t],1));
```
Comment (alaparicioc) [inline src/2_model/alt/models_doug/2_props_model/20_props_model.stan]: I got the following error "rray[uni, ...] index: accessing element out of range. index 80 out of range; expecting index to be between 1 and 79 (in '/tmp/Rtmpp7N6Zq/model-eff4455534a.stan', line 97, column 4 to column 47)". Imay be defining something wrong

**Review by alaparicioc — CHANGES_REQUESTED** [review]
Hi,
This has been great fun to work with this models and comparing it with what I have tried.

Regarding the 2_props_model.stan and 20_props_model.stan models. I think Nick did a great job optimising the model, particularly in improving the early-period estimates by using the logit-transformed version (logit_pi0) as the prior mean for logit_pi[1]. This seems to provide a stronger empirical prior for initial proportions, reflecting the initial population, reducing uncertainty, and decreasing the number of parameters to estimate, and threfore time to estimate. However, I got the following message:
image
There is divergence in the last date defined in the 10_create_master_index.R and the md created when running the mcmc.R for the 20_props_mod.

In addition, if more flexibility in the initialisation of pi[1] is needed, or if greater flexibility in early estimates of proportions is preferred, then keeping pi[1] via softmax(logit_pi[1]) while retaining N0 could be a good alternative. In addition, in this model the treedepth is nice.

Regarding 3_covs_model.stan, I would lean towards using the lognormal distribution for log(y_F) and log(y_G), as I have found it to perform better in the models I have run.

I also wanted to ask whether you have tested the negative binomial lognormal observation model you mentioned at some point. Were you referring to applying the lognormal distribution to p_F and p_G, or to the full expression N * p_F? I’d be interested in hearing your insights.

Additionally, I am still curious about what you learned regarding the E-BFMI issue—was it related to initialisation, parameter scaling, or something else? I am still encountering it.

Lastly, you mentioned the undercoverage of prediction intervals when using only a Poisson model (and a negative binomial?) in tests with simulated data. I’d be keen to hear more about what you found.

Andrea

File: src/2_model/alt/models_doug/2_props_model/20_props_model.stan [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+}
+model {
+  // log-likelihood
+  target += sum(log_lik);
+
+  // empirical priors
+  N[ti_N1] ~ lognormal(log(N1), ci_N1 / 2);
+
+  // population growth rates
+  logit_pi[1] ~ normal(logit_pi0, exp(log_sigma_pi));
```
Comment (alaparicioc) [inline src/2_model/alt/models_doug/2_props_model/20_props_model.stan]: I am assuming that you were thinking that covariates for helping estimating changes in proportions could be here. Something like?:
for (t in 2:T) {
    vector[I-1] mu_logit_pi = logit_pi[t-1] + X_pi[t_slice(t, I), ] * beta_pi;
    logit_pi[t] ~ normal(mu_logit_pi, exp(log_sigma_pi));
  }

File: src/2_model/alt/models_doug/2_props_model/20_props_model.stan (hunk 1 above)
Comment (doug-leasure) [reply src/2_model/alt/models_doug/2_props_model/20_props_model.stan]: Fixed with commit 308c1bcde5ce9423879f6d75e50dd86d9dca1b0e

**Comment by doug-leasure** [comment]
Thanks, Andrea! Here are responses to you points.

> However, I got the following message: image 

The indexing issue is now fixed with commit 308c1bcde5ce9423879f6d75e50dd86d9dca1b0e.

> There is divergence in the last date defined in the 10_create_master_index.R and the md created when running the mcmc.R for the 20_props_mod.
 
The last date in the `md` object is defined using the `model_data()` function here: https://github.com/OxfordDemSci/UkraineNowPop/blob/68-model-with-covariates/src/2_model/1_mcmc.R#L58. It does not necessarily need to correspond to the last date in the master index.

> In addition, if more flexibility in the initialisation of pi[1] is needed, or if greater flexibility in early estimates of proportions is preferred, then keeping pi[1] via softmax(logit_pi[1]) while retaining N0 could be a good alternative. 

I think the model is initialising okay, but please let me know if you are seeing problems with initialisation now that the indexing issue has been resolved so that you can run the model. 

> In addition, in this model the treedepth is nice.

The tree depth warnings are now improved for both models, although still better for the props model. Note that the tree depth warning is only an indication of sampler efficiency (see [here](https://github.com/OxfordDemSci/UkraineNowPop/blob/68-model-with-covariates/src/2_model/1_mcmc.R#L58)) rather than a issue with validity of inferences. I'm not too worried about this warning if the models are converging nicely--which they seem to be. Were you having any issues with convergence or running time? 
 
> Regarding 3_covs_model.stan, I would lean towards using the lognormal distribution for log(y_F) and log(y_G), as I have found it to perform better in the models I have run.

This model is using lognormal likelihoods for `y_F` and `y_G`. See here: https://github.com/OxfordDemSci/UkraineNowPop/blob/68-model-with-covariates/src/2_model/models/3_covs_model.stan#L116. 

Is your suggestion to use `log(y_F)` as the outcome rather than `y_F`?
 
> I also wanted to ask whether you have tested the negative binomial lognormal observation model you mentioned at some point. Were you referring to applying the lognormal distribution to p_F and p_G, or to the full expression N * p_F? I’d be interested in hearing your insights.

I didn't do much testing of the negative binomial other than running Nick's models and assessing their performance (which was good). Here, I chose to use a lognormal likelihood with the expected value of `log(N * p_F)`. 

https://github.com/OxfordDemSci/UkraineNowPop/blob/68-model-with-covariates/src/2_model/models/3_covs_model.stan#L116

> Additionally, I am still curious about what you learned regarding the E-BFMI issue—was it related to initialisation, parameter scaling, or something else? I am still encountering it.

I found through simulating data that E-BFMI is more likely to be NA when the likelihood is discrete and y_F has large numbers (e.g. tens of thousands). There may be other factors at play, but this is one driver that was replicable. 

> Lastly, you mentioned the undercoverage of prediction intervals when using only a Poisson model (and a negative binomial?) in tests with simulated data. I’d be keen to hear more about what you found.

There was undercoverage using the Poisson_logNormal likelihood with real data. Nick's negative binomial-logNormal likelihood resolved that issue (see his results in PR #93). 

The current models in this PR use a logNormal likelihood (no negative binomial) and gets good coverage of prediction intervals.

File: src/2_model/alt/models_doug/2_props_model/20_props_model.stan (hunk 2 above)
Comment (doug-leasure) [reply src/2_model/alt/models_doug/2_props_model/20_props_model.stan]: Yep, that is the idea exactly. There are a few little complications (e.g. in this formulation, the regression is omitting one location because logit_pi is of length I-1), but should be resolvable. 

Another idea was to derive the population growth rate as a transformed parameter and put the regression on that parameter.

**Comment by alaparicioc** [comment]
> The indexing issue is now fixed with commit [308c1bc](https://github.com/OxfordDemSci/UkraineNowPop/commit/308c1bcde5ce9423879f6d75e50dd86d9dca1b0e).

Yes, I could run the code.
 
> I think the model is initialising okay, but please let me know if you are seeing problems with initialisation now that the indexing issue has been resolved so that you can run the model.

Even though both models (2_propos_model and 3_covs_model) initiated, there is divergence in the estimated population sizes (i.e. the CIs do not overlap). As expected, the divergences are for the Autonomous and East oblasts. I was surprised that tthe city also got some difference. 
image
 
> > In addition, in this model the treedepth is nice.
> 
> The tree depth warnings are now improved for both models, although still better for the props model. Note that the tree depth warning is only an indication of sampler efficiency (see [here](https://github.com/OxfordDemSci/UkraineNowPop/blob/68-model-with-covariates/src/2_model/1_mcmc.R#L58)) rather than a issue with validity of inferences. I'm not too worried about this warning if the models are converging nicely--which they seem to be. Were you having any issues with convergence or running time?

With the initial 1000 itearations, yes, especially with the term you named as logit_pi_free.

**Review by alaparicioc — APPROVED** [review]

**Comment by doug-leasure** [comment]
> Even though both models (2_propos_model and 3_covs_model) initiated, there is divergence in the estimated population sizes (i.e. the CIs do not overlap). As expected, the divergences are for the Autonomous and East oblasts. I was surprised that tthe city also got some difference. 

This is nice to see. Thanks for looking at it in this way, @alaparicioc! It will be interesting to see if the population estimates still differ between models once we incorporate covariates into the proportions model. Those oblasts that you highlight have no data for `y_F` and `y_G`, so their estimates are currently limited to information from the random walk in the props process model.
